packages/loopring_v3/contracts/amm/AmplifiedAmmController.sol

```diff
@@ -38,24 +43,41 @@ contract AmplifiedAmmController is IAmmController, Claimable
         return vTokenBalancesL2;
     }
 
-    function getVirtualBalances(
-        uint96[] memory tokenBalancesL2,
-        uint96[] memory /*vTokenBalancesL2*/
+   function authorizeVirtualBalances(
+        uint96[] memory balances,
+        uint96[] memory /*vBalancesOld*/,
+        uint96[] memory vBalancesNew,
+        bytes    memory /*data*/
         )
         external
-        view
         override
-        returns (uint96[] memory)
+        returns (bool)
     {
-        // Only allow updating the virtual balances if the AF = 1
-        require(getAmplificationFactor(msg.sender) == AMPLIFICATION_FACTOR_BASE, "INVALID_OPERATION");
+        uint authorizedTimestamp = curveChangeAuthorization[msg.sender];
+        bool authorized = (authorizedTimestamp > 0) && block.timestamp >= curveChangeAuthorization[msg.sender];
+
+        // Remove authorization
+        delete curveChangeAuthorization[msg.sender];
 
-        // Just set the virtual balances to the actual balances
-        uint96[] memory vTokenBalancesL2 = new uint96[](tokenBalancesL2.length);
-        for (uint i = 0; i < tokenBalancesL2.length; i++) {
-            vTokenBalancesL2[i] = tokenBalancesL2[i];
+        // Special case: Always allow updating the virtual balances if the AF = 1
+        if (!authorized && getAmplificationFactor(msg.sender) == AMPLIFICATION_FACTOR_BASE) {
```

Comment:
if `authorized==true`, this method will return false, which doesn't seem right to me.